### ISSUE TITLE
Instantiation and State Isolation

### ISSUE THREAD

**Issue by tomkis1** [erikras/ducks-modular-redux#20]
> To me, it makes more sense for these components to be bundled together in an isolated module that is self contained, and can even be packaged easily into a library.

I believe that ducks is definitely a great pattern for Component modularization. However, a typical use case for bundled Components is instantiation. Let's say that we are building a Text Field. We want to have many instances of Text Field on the page, each with its internal state. There have been countless discussion about [local state in redux](https://github.com/rackt/redux/issues/159) yet with no specific outcome.

I can imagine that we could have one or two "duck" files where one would contain Root Component Reducer and all corresponding Actions and the other would contain corresponding View (which may or may not be a React Component).

> Java has jars and beans. Ruby has gems. I suggest we call these reducer bundles "ducks", as in the last syllable of "redux".

Java has beans but beans are just Classes not Instances. The UI component need instantiation because of the internal state. So we need to figure out the way how to isolate state for the specific view component.

I believe that every duck action should contain `duckId` which is actually a form of correlation ID responsible for grouping all the related actions together for specific component instance. Then it's just a matter of implementing reducer enhancer which is responsible for pairing the actions with specific state slice and providing it to corresponding reducer. By doing this, we can simply achieve completely isolated instantiable UI Components which are very easy to publish as a library.

I believe that `ducks-modular-redux` can help us with solving really hard problem in Redux and it's Component pseudo local state (cc @gaearon).

I did two repositories (proof of concept) which tackle the problem above.
- [https://github.com/salsita/redux-ducks](redux-ducks) for state isolation
- [https://github.com/salsita/react-redux-ducks](react-redux-ducks) which is basically a very thin React binding so that we are view independent

I would like to know your opinions.

**Comment by erikras** [comment]
I was looking at this earlier when you @ mentioned me on Twitter. It reminds me a bit of my [multireducer](https://github.com/erikras/multireducer) lib (also mostly a proof of concept).

**Comment by tomkis1** [comment]
Yes, there are indeed few similarities and I believe this kind of libraries will gain more popularity with concepts like `ducks-modular-redux` as it's essential to figure out right way how to treat pseudo local component state.

**Comment by kuon** [comment]
The newest library I found, https://github.com/gcazaciuc/redux-fractal has a nice approach, but I have a feeling this can still be improved.

I'd like to see this feature appears in https://github.com/reactjs/react-redux seems similar to `connect` in nature to me.

So far I found:

https://github.com/salsita/react-redux-ducks
https://github.com/gcazaciuc/redux-fractal
https://github.com/Dashlane/redux-cursor
https://github.com/lastmjs/redux-store-element
https://github.com/threepointone/redux-react-local

There have been countless discussions:

https://github.com/reactjs/react-redux/issues/275
https://github.com/reactjs/redux/issues/159

I really think this problem should be "tackled" in an organized way, there are a dozen of repositories with different approach to it, and I don't think this fragmentation is good for the community.

Until now, this has been "dodged" multiple times while being acknowledged as being required.

This might not be the best place to discuss it, but after two days of turning around the problem, I came to a stop as, again, it feels like this should be included in https://github.com/reactjs/react-redux

cc @gaearon @denis-sokolov @adriancooney

**Comment by denis-sokolov** [comment]
I think we'll evolve into a "best" solution once something catches on on its own merits.

**Comment by kuon** [comment]
Yeah, maybe the different "experimentation" are a good thing to find the best way to do it.

Anyway, I tried all the approaches above, and <https://github.com/gcazaciuc/redux-fractal> is the first one fitting all my needs/paradigm I'd like to use.
